src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.Buffer.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System.Diagnostics;
+using System.Security;
+
+namespace System.Numerics
+{
+    internal static partial class BigIntegerCalculator
+    {
+        // To spare memory allocations a buffer helps reusing memory!
+        // We just create the target array twice and switch between every
+        // operation. In order to not compute unnecessarily with all those
+        // leading zeros we take care of the current actual length.
+
+        // ATTENTION: always pass this by reference,
+        // it's a structure for performance reasons
+
+        internal struct BitsBuffer
+        {
+            private uint[] _bits;
+            private int _length;
+
+            public BitsBuffer(int length, uint value)
+            {
+                Debug.Assert(length >= 1);
+
+                _bits = new uint[length];
+                _length = value != 0 ? 1 : 0;
+
+                _bits[0] = value;
+            }
+
+            public BitsBuffer(int length, uint[] value)
+            {
+                Debug.Assert(value != null);
+                Debug.Assert(length >= ActualLength(value));
+
+                _bits = new uint[length];
+                _length = ActualLength(value);
+
+                Array.Copy(value, _bits, _length);
```

Comment:
```Array.Copy(value, 0, _bits, 0, _length);``` helps to save having to query for the lower bounds of the arrays, which can be non-zero with Array.